### PR TITLE
Introduce Kubernetes as a cloud provider in clouddriver

### PR DESCRIPTION
**Pull request by lwander**
@duftler @ttomsu @skim1420

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -77,6 +77,7 @@ subprojects { project ->
 
   dependencies {
     compile spinnaker.dependency('groovy')
+    compile 'io.fabric8:kubernetes-api:2.2.70'
```
Comment (duftler) [inline build.gradle]: Drop a todo here to move this here: https://github.com/spinnaker/spinnaker-dependencies/blob/master/src/spinnaker-dependencies.yml

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/KubernetesConfiguration.groovy [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    new KubernetesConfigurationProperties()
+  }
+
+  @Bean
+  KubernetesHealthIndicator kubernetesHealthIndicator() {
+    new KubernetesHealthIndicator()
+  }
+
+  @Bean
+  String kubernetesApplicationName(@Value('${Implementation-Version:Unknown}') String implementationVersion) {
```
Comment (duftler) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/KubernetesConfiguration.groovy]: Probably don't need this (at least not yet). This value in the GCE integration is mainly used for UA headers. Are you using it for something similar on k8s calls.

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/KubernetesConfiguration.groovy [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+import org.springframework.context.annotation.PropertySource
+import org.springframework.context.annotation.Scope
+import org.springframework.scheduling.annotation.EnableScheduling
+
+@Configuration
+@EnableConfigurationProperties
+@EnableScheduling
+@ConditionalOnProperty('kubernetes.enabled')
+@ComponentScan(["com.netflix.spinnaker.clouddriver.kubernetes"])
+@PropertySource(value = "classpath:META-INF/clouddriver-core.properties", ignoreResourceNotFound = true)
```
Comment (duftler) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/KubernetesConfiguration.groovy]: If you decide you don't need `kubernetesApplicationName`, you probably won't need this either.

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/config/KubernetesConfigurationProperties.groovy [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.clouddriver.kubernetes.config
+
+import groovy.transform.ToString
+
+@ToString(includeNames = true)
+class KubernetesConfigurationProperties {
+  public static final int POLLING_INTERVAL_SECONDS_DEFAULT = 60
```
Comment (duftler) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/config/KubernetesConfigurationProperties.groovy]: Don't think these are needed until we start surfacing the k8s resources.

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/security/KubernetesCredentialsInitializer.groovy [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+        accountCredentialsRepository.save(managedAccount.name, kubernetesAccount)
+      } catch (e) {
+        log.info "Could not load account ${managedAccount.name} for Kubernetes.", e
+      }
+    }
+
+    ProviderUtils.unscheduleAndDeregisterAgents(namesOfDeletedAccounts, catsModule)
+
+    /*
```
Comment (duftler) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/security/KubernetesCredentialsInitializer.groovy]: Please add a comment as to why this is commented-out, but not remove entirely. (I'm ok leaving it, just drop a comment.)

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/security/KubernetesCredentialsInitializer.groovy [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+  List<?> synchronizeKubernetesAccounts(KubernetesConfigurationProperties kubernetesConfigurationProperties, CatsModule catsModule) {
+    def (ArrayList<KubernetesConfigurationProperties.ManagedAccount> accountsToAdd, List<String> namesOfDeletedAccounts) =
+      ProviderUtils.calculateAccountDeltas(accountCredentialsRepository,
+                                           KubernetesNamedAccountCredentials,
+                                           kubernetesConfigurationProperties.accounts)
+
+    accountsToAdd.each { KubernetesConfigurationProperties.ManagedAccount managedAccount ->
+      try {
+        def kubernetesAccount = new KubernetesNamedAccountCredentials(managedAccount.name,
+                                                        managedAccount.environment ?: managedAccount.name,
```
Comment (duftler) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/security/KubernetesCredentialsInitializer.groovy]: Formatting is off here.

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/health/KubernetesHealthIndicator.groovy [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+  void checkHealth() {
+    try {
+      Set<KubernetesNamedAccountCredentials> kubernetesCredentialsSet = accountCredentialsProvider.all.findAll {
+        it instanceof KubernetesNamedAccountCredentials
+      } as Set<KubernetesNamedAccountCredentials>
+
+      for (KubernetesNamedAccountCredentials accountCredentials in kubernetesCredentialsSet) {
+        KubernetesCredentials kubernetesCredentials = accountCredentials.credentials
+
+        // This verifies that the specified credentials are sufficient to access the referenced project.
```
Comment (duftler) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/health/KubernetesHealthIndicator.groovy]: Not a "project", but a k8s cluster/namespace, right?

Comment (duftler) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/health/KubernetesHealthIndicator.groovy]: Or master/namespace...

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/KubernetesConfiguration.groovy (hunk 2 above)
Comment (lwander) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/KubernetesConfiguration.groovy]: Is it bad to leave the UA in? I don't think Kubernetes logs it, but someone else might be monitoring their own network traffic.

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/config/KubernetesConfigurationProperties.groovy (hunk 4 above)
Comment (lwander) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/config/KubernetesConfigurationProperties.groovy]: Should I comment them out? I figured I'd leave them so I don't have to concurrently make changes between Deck and Clouddriver.

Comment (duftler) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/config/KubernetesConfigurationProperties.groovy]: I'd lean towards removing them (and the 3 associated fields) until they are necessary.

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/KubernetesConfiguration.groovy (hunk 2 above)
Comment (duftler) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/KubernetesConfiguration.groovy]: They would only be used for UA headers if you explicitly set them on the kubernetes client calls, as we do on the GCE client calls.

Comment (lwander) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/KubernetesConfiguration.groovy]: Oh I see, then these don't need to be here.

**Comment by duftler** [comment]
LGTM.

**Pull request merged**